File: backend/app/domain/documents/schemas/vault.py
"""Document vault payloads: the category tree, file records, and version history."""

from __future__ import annotations

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from app.domain.documents.models.enums import DocumentKind, DocumentVisibility


class CategoryNode(BaseModel):
    """A node in the filing tree, carrying its own children so one request renders the sidebar."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    slug: str
    name_en: str
    name_zh: str
    description: str | None = None
    kind: DocumentKind
    department_id: UUID | None = None
    parent_id: UUID | None = None
    is_system_managed: bool
    sort_order: int
    document_count: int = 0
    children: list[CategoryNode] = Field(default_factory=list)


class CategoryCreate(BaseModel):
    slug: str = Field(min_length=1, max_length=80, pattern="^[a-z0-9][a-z0-9-]*$")
    name_en: str = Field(min_length=1, max_length=160)
    name_zh: str = Field(min_length=1, max_length=160)
    description: str | None = Field(default=None, max_length=1000)
    kind: DocumentKind = DocumentKind.GOVERNANCE
    department_id: UUID | None = None
    parent_id: UUID | None = None
    sort_order: int = 0


class CategoryUpdate(BaseModel):
    name_en: str | None = Field(default=None, min_length=1, max_length=160)
    name_zh: str | None = Field(default=None, min_length=1, max_length=160)
    description: str | None = Field(default=None, max_length=1000)
    kind: DocumentKind | None = None
    parent_id: UUID | None = None
    sort_order: int | None = None


class DocumentVersionSummary(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    version_number: int
    original_filename: str
    content_type: str
    size_bytes: int
    checksum_sha256: str
    change_note: str | None = None
    uploaded_by_id: UUID | None = None
    uploaded_by_name: str | None = None
    created_at: datetime


class DocumentSummary(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    title: str
    description: str | None = None
    category_id: UUID
    category_name_zh: str | None = None
    department_id: UUID
    department_slug: str | None = None
    department_name_zh: str | None = None
    visibility: DocumentVisibility
    tags: list[str] = Field(default_factory=list)
    is_archived: bool = False
    latest_version: DocumentVersionSummary | None = None
    updated_at: datetime


class DocumentDetail(DocumentSummary):
    versions: list[DocumentVersionSummary] = Field(default_factory=list)
    uploaded_by_id: UUID | None = None
    uploaded_by_name: str | None = None


class DocumentPage(BaseModel):
    items: list[DocumentSummary]
    total: int
    page: int
    page_size: int


class DocumentUpdate(BaseModel):
    title: str | None = Field(default=None, min_length=1, max_length=240)
    description: str | None = Field(default=None, max_length=2000)
    category_id: UUID | None = None
    visibility: DocumentVisibility | None = None
    tags: list[str] | None = None
    is_archived: bool | None = None


class DownloadTicket(BaseModel):
    """Short-lived location a browser can fetch the bytes from.

    Object storage returns a presigned URL; the local driver streams through the API instead, in
    which case ``url`` points back at this service.

    对象存储返回预签名 URL；本地驱动则由 API 直接流式返回，此时 url 指向本服务自身。
    """

    url: str
    filename: str
    content_type: str
    size_bytes: int
    expires_at: datetime | None = None

File: frontend/src/shared/api/contracts/documents/vault.ts
import type { PaginatedCollection } from '@/shared/api/contracts/common/pagination'
import type { IsoDateTimeString, UuidString } from '@/shared/api/contracts/common/scalarTypes'
import type {
  DocumentKind,
  DocumentVisibility,
} from '@/shared/api/contracts/documents/documentClassification'

/** A node in the filing tree that carries its own children, so one request renders the sidebar. */
export interface CategoryNode {
  id: UuidString
  slug: string
  name_en: string
  name_zh: string
  description?: string | null
  kind: DocumentKind
  department_id?: UuidString | null
  parent_id?: UuidString | null
  is_system_managed: boolean
  sort_order: number
  document_count: number
  children: CategoryNode[]
}

export interface CategoryCreate {
  slug: string
  name_en: string
  name_zh: string
  description?: string | null
  kind: DocumentKind
  department_id?: UuidString | null
  parent_id?: UuidString | null
  sort_order: number
}

export interface CategoryUpdate {
  name_en?: string | null
  name_zh?: string | null
  description?: string | null
  kind?: DocumentKind | null
  parent_id?: UuidString | null
  sort_order?: number | null
}

export interface DocumentVersionSummary {
  id: UuidString
  version_number: number
  original_filename: string
  content_type: string
  size_bytes: number
  checksum_sha256: string
  change_note?: string | null
  uploaded_by_id?: UuidString | null
  uploaded_by_name?: string | null
  created_at: IsoDateTimeString
}

export interface DocumentSummary {
  id: UuidString
  title: string
  description?: string | null
  category_id: UuidString
  category_name_zh?: string | null
  department_id: UuidString
  department_slug?: string | null
  department_name_zh?: string | null
  visibility: DocumentVisibility
  tags: string[]
  is_archived: boolean
  latest_version?: DocumentVersionSummary | null
  updated_at: IsoDateTimeString
}

export interface DocumentDetail extends DocumentSummary {
  versions: DocumentVersionSummary[]
  uploaded_by_id?: UuidString | null
  uploaded_by_name?: string | null
}

export type DocumentPage = PaginatedCollection<DocumentSummary>

export interface DocumentUpdate {
  title?: string | null
  description?: string | null
  category_id?: UuidString | null
  visibility?: DocumentVisibility | null
  tags?: string[] | null
  is_archived?: boolean | null
}

/**
 * Short-lived location the browser fetches the bytes from; object storage returns a presigned URL
 * while the local driver points back at the API itself.
 *
 * 浏览器据此获取文件字节的短时地址：对象存储返回预签名 URL，本地驱动则指向 API 自身。
 */
export interface DownloadTicket {
  url: string
  filename: string
  content_type: string
  size_bytes: number
  expires_at?: IsoDateTimeString | null
}

export interface DocumentQuery {
  department_id?: UuidString
  category_id?: UuidString
  search?: string
  page?: number
  page_size?: number
}

/** Multipart upload fields; `tags` is sent as a repeated form field. */
export interface DocumentUploadInput {
  file: File
  title: string
  description?: string
  category_id: UuidString
  department_id: UuidString
  visibility: DocumentVisibility
  tags: string[]
}

export interface DocumentVersionUploadInput {
  file: File
  change_note?: string
}
